AWSAppSyncClient/AWSAppSyncHTTPNetworkTransport.swift

```diff
@@ -249,9 +250,13 @@ public class AWSAppSyncHTTPNetworkTransport: AWSNetworkTransport {
                                                          retryHandler: retryHandler,
                                                          networkTransportOperation: networkTransportOperation,
                                                          completionHandler: completionHandler)
-                                self?.activeTimers.removeValue(forKey: taskUUID)
+                                self?.activeTimersQueue.async(flags: .barrier) {
```

Comment:
What I find works best is to have private functions which handle protected state transitions which you can sync. With private functions you know there is no way to call those functions from outside this type, so you can strictly control the context of the work. I like to use `dispatchPrecondition` to ensure a private function is not on the internal queue before calling `queue.sync` in it. And I want to use `sync` instead of `async` if it is just a small task like changing the value of a stored property. For a longer running task I would want to make sure it is run off the `.main` queue by moving the work to `global()` and then syncing on the internal queue. You never want to block the main thread for too long as it risk the watchdog killing the app.

With private functions you can also ensure these functions to do not call each other within the sync block. And if you only use the queue in these private functions and have `dispatchPrecondition` guarding the context when entering the functions you will make sure you cannot reach a deadlock. 

When you use `sync` you also know the stored property is set by the time execution reaches the next line.

Reply:
I've prepared this [PR](https://github.com/awslabs/aws-mobile-appsync-sdk-ios/pull/503) with these changes on top of your changes. The unit tests are passing.

Reply:
thanks, merged https://github.com/awslabs/aws-mobile-appsync-sdk-ios/pull/503 into this branch